### ISSUE TITLE
Vmagent: is stuck on scraping when multiple big storages are unavailable

### ISSUE THREAD

**Issue by Vladikamira** [VictoriaMetrics/VictoriaMetrics#2802]
**Describe the bug**
We collect all our metrics from exporters at each k8s cluster by 2 replicas of `vmagent`. It points to 3 VM clusters (1 per zone), so we have 3 copy of data. It works pretty stable until we need to restart/redeploy clusters one by one with some delay. 
So, on some big clusters, when more than 1 zone is unavailable to write data, we see how `vmagent` is hanging, consuming less CPU, not scraping all targets, and consuming a lot more memory. You can see that from the Screenshot 
![Screenshot 2022-06-30 at 12 09 03](https://user-images.githubusercontent.com/5423271/176655896-0f917401-a8e5-4b1e-aba7-a04756cf42a6.png)
I also tried to capture `pprof` from one of the agents at the problem time, but not sure that I made it correct, please have a look
[pprof_vmagent.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/9018952/pprof_vmagent.zip)

**To Reproduce**
Probably need to play with restarting of remote-write backends when scraping more than 1-2 million samples

**Expected behavior**
`Vmagent` keep collecting scrapped metrics from targets until the storage will be available for write

**Version**
```
/ # /vmagent-prod --version
vmagent-20220318-180512-tags-v1.75.0-0-g87d356348
```

**Used command-line flags**
```
      -promscrape.config=/config/scrape.yml
      -remoteWrite.tmpDataPath=/tmpData
      -remoteWrite.url=http://victoria-metrics-cluster-a-vminsert.monitoring.svc.cluster.local:8480/insert/0/prometheus/
      -remoteWrite.url=http://victoria-metrics-cluster-b-vminsert.monitoring.svc.cluster.local:8480/insert/0/prometheus/
      -remoteWrite.url=http://victoria-metrics-cluster-c-vminsert.monitoring.svc.cluster.local:8480/insert/0/prometheus/
      -envflag.enable=true
      -envflag.prefix=VM_
      -loggerErrorsPerSecondLimit=2
      -loggerFormat=json
      -loggerLevel=WARN
      -loggerOutput=stdout
      -loggerWarnsPerSecondLimit=1
      -memory.allowedPercent=60
      -promscrape.config=/config/prometheus.yml
      -promscrape.config.strictParse=false
      -promscrape.configCheckInterval=10s
      -promscrape.maxScrapeSize=1024MB
      -promscrape.suppressScrapeErrors=true
      -remoteWrite.showURL=true
      -remoteWrite.queues=60
```

**Comment by Vladikamira** [comment]
right.. good point 😅 not all of them are lost obviously, as I showed in the graph above metrics from `Vmagent` itself
so, it some part, have a look
![Screenshot 2022-06-30 at 13 51 07](https://user-images.githubusercontent.com/5423271/176670736-8f8f3e4a-814b-4240-98a7-19e452e29bb5.png)

I can try to find specific metrics names if it will help

**Comment by hagen1778** [comment]
Hi @Vladikamira!

Can I ask you to plot two following queries during the incident?
```
sum(rate(vm_persistentqueue_write_duration_seconds_total{job=~"$vmagent_job"})) by (instance)
sum(rate(vm_persistentqueue_read_duration_seconds_total{job=~"$vmagent_job"})) by (instance)
```

**Comment by Vladikamira** [comment]
sure, here
<img width="1439" alt="Screenshot 2022-07-01 at 14 37 02" src="https://user-images.githubusercontent.com/5423271/176895899-d605caf5-9050-4c22-825e-a62819d8a9c0.png">
and here
<img width="1439" alt="Screenshot 2022-07-01 at 14 37 21" src="https://user-images.githubusercontent.com/5423271/176895950-6ce43934-d2e6-4450-9280-97423a8ed18f.png">

**Comment by hagen1778** [comment]
Thanks! I think degradation of samples scraping is caused by disk saturation for vmagent.

For each target vmagent runs a separate goroutine. This goroutine periodically runs a GET request, receives the response, parses it and pushes results to the remote-write queue.

Vmagent maintains in-memory buffers for remote-write queues. Once there is a data in queue, reader consumes it, forms a request and sends to remote storage via remote-write connection.

During scraping, each goroutine is blocked by target response time (limited by deadline), response processing time and time needed to push to the remote-write queue. When in-memory buffers are empty, adding data to buffer is a relatively fast operation. But when buffer is full, all excessive pushes to the queue end up flushing data to the disk - to the persistent queue.

Metrics  `vm_persistentqueue_write_duration_seconds_total ` and `vm_persistentqueue_read_duration_seconds_total ` show how much time was spent on writing or reading from the queue respectively. Rate over these metrics shows relative **saturation** of these calls. If rate is > 1s, it means all the time was spent just waiting for queue to respond or that saturation is 100%. When time spent on pushing to the persistent queue exceeds scrape interval, goroutines start to miss scrapes.

I think, this is what we see on the provided screenshots. When vmagent starts to buffer data on disk or read from it, the main resources it can be limited by are disk read/write throughput and IOPS limit. Is there a way how you can check that there was enough of throughput or IOPS for vmagent during the incident?

If these limits are a reason of vmagent degradation, then possible workarounds are the following:
* increase throughput and IOPS limits for vmagent's disks
* shard vmagents - see https://docs.victoriametrics.com/vmagent.html#scraping-big-number-of-targets

**Comment by Vladikamira** [comment]
oh, that explains everything! 
because we are using 300Gb PD Standard GCP disks on each `VMagent` which according to [GCP Spec](https://cloud.google.com/compute/docs/disks#pdspecs) gives us these numbers:
```
IOPS          1.5 * 300  = 450 iops
Throughput    0.12 * 300 = 36 MB/s
```
I don't know how to measure/calculate utilization of GCP disks, but I'm pretty sure it is the cause here, as it only happens on the two most significant clusters. I will increase disk capacity and try to redeploy next week to confirm that it solves the problem.

**Comment by Vladikamira** [comment]
I can confirm that increasing IOps/Throughput on disk for the Persistent queue of `Vmagent` solved the problem!
Thanks a lot @hagen1778 for your detailed explanation! I think we can close this issue
